Java: Implement DatabaseMetaData#isReadOnly()

MariusVolkhart opened the pull request:
Delegate to the connection.

Mytherin commented:
Thanks! Looks good. The windows test seems to fail due to a file locking issue, unfortunately. I believe that is why this test was previously commented out. I suppose we could add a sleep or maybe an explicit GC activation to try and fix this?

MariusVolkhart committed:
JDBC: Fix read_only test

The assertion were broken, since fail() throws Exception, which was immediately being swallowed. It turns out, we CAN open multiple connections.

During CI, we were still seeing errors on Windows because the database file was still held open. That's because the Connection created by the driver still held the file. It wasn't that we needed GC to run so much as it was that the assumptions of the test didn't match reality.

MariusVolkhart committed:
JDBC: Implement DatabaseMetaData#isReadOnly()

Delegate to the connection.

MariusVolkhart commented on tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:
```diff
@@ -1476,76 +1474,78 @@ public static void test_prepare_insert() throws Exception {
-			fail();
-		} catch (Exception e) {
-		}
+		// Verify we can open additional write connections
```
Interestingly, opening additional connections isn't a problem. I think we need to update the documentation (website). There's only 1 scenario that I was able to find that didn't work: mixing RW and RO connections, see line 1507.

MariusVolkhart commented on tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:
```diff
@@ -1476,76 +1474,78 @@ public static void test_prepare_insert() throws Exception {
-		try {
-			Connection conn2 = DriverManager.getConnection(jdbc_url);
-			fail();
-		} catch (Exception e) {
```
This block of code was actually what was causing this test to fail on Windows. The `conn2` opens successfully. However, we then hit the `fail()`, which throws, but is immediately caught by the `catch`, hiding the problem.

Since the connection opened, there's an additional handle to the database file open. Using `lsof` on Mac, I was able to see that prior to the changes in this PR, Java kept the DB file open after this block, and after the changes, it doesn't.

MariusVolkhart commented on tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:
```diff
@@ -1476,76 +1474,78 @@ public static void test_prepare_insert() throws Exception {
-		// hard shutdown to not have to wait on gc
-		DuckDBDatabase db = ((DuckDBConnection) conn_rw).getDatabase();
 		conn_rw.close();
-		db.shutdown();
```
Closing the last connection closes the database. We have a separate test verifying this behavior.

MariusVolkhart reviewed (commented):
The most recent rebase splits the changes into 2 commits. The first modifies the test so it 1. works on Windows and 2. isn't broken. The second commit adds the new functionality and assertions for it.
